cpp/src/interop/arrow_data_structures.cpp

```diff
@@ -0,0 +1,345 @@
+  std::vector<std::unique_ptr<ArrowArray>> children;
+  std::vector<ArrowArray*> children_raw;
+};
+
+/**
+ * @brief Release callback for an ArrowArray that contains a struct array.
+ *
+ * This function is called when the ArrowArray is released. It releases all of the children of the
+ * struct array.
+ *
+ * @param array The ArrowArray to release
+ */
+void ArrayReleaseCallback(ArrowArray* array)
+{
+  auto private_data = reinterpret_cast<ArrowArrayPrivateData*>(array->private_data);
+  for (auto& child : private_data->children) {
+    child->release(child.get());
+  }
+  delete private_data;
+  array->release = nullptr;
+}
+
+/**
+ * @brief Copy an ArrowArray.
+ *
+ * This function shallow copies an ArrowArray and all of its children. It is
+ * used to export cudf arrow objects to user-provided ArrowDeviceArrays.
+ *
+ * @param output The ArrowArray to copy to
+ * @param input The ArrowArray to copy from
+ * @param container The container that owns the data
+ */
+void copy_array(ArrowArray* output,
+                ArrowArray const* input,
+                std::shared_ptr<arrow_array_container> container)
+{
+  auto private_data  = new ArrowArrayPrivateData{container};
+  output->length     = input->length;
+  output->null_count = input->null_count;
+  output->offset     = input->offset;
+  output->n_buffers  = input->n_buffers;
+  output->n_children = input->n_children;
+  output->buffers    = input->buffers;
+
+  if (input->n_children > 0) {
+    private_data->children_raw.resize(input->n_children);
+    for (auto i = 0; i < input->n_children; ++i) {
+      private_data->children.push_back(std::make_unique<ArrowArray>());
+      private_data->children_raw[i] = private_data->children.back().get();
+      copy_array(private_data->children_raw[i], input->children[i], container);
+    }
+  }
+  output->children     = private_data->children_raw.data();
+  output->dictionary   = input->dictionary;
+  output->release      = ArrayReleaseCallback;
+  output->private_data = private_data;
+}
+
+template <typename T>
+void arrow_obj_to_arrow(T& obj,
+                        std::shared_ptr<arrow_array_container> container,
+                        ArrowDeviceArray* output,
+                        ArrowDeviceType device_type,
+                        rmm::cuda_stream_view stream,
+                        rmm::device_async_resource_ref mr)
+{
+  switch (device_type) {
+    case ARROW_DEVICE_CUDA:
+    case ARROW_DEVICE_CUDA_HOST:
+    case ARROW_DEVICE_CUDA_MANAGED: {
+      auto& device_arr = container->owner;
+      copy_array(&output->array, &device_arr.array, container);
+      output->device_id = device_arr.device_id;
+      // We can reuse the sync event by reference from the input. The
+      // destruction of that event is managed by the destruction of
+      // the underlying ArrowDeviceArray of this table.
+      output->sync_event  = device_arr.sync_event;
+      output->device_type = device_type;
+      break;
```

Comment:
if `output->release` gets called, now both `output` and the `container->owner` array will be released, only `container->owner` won't know this and will still have a valid `release` callback function pointer. That seems like a problem, right?

Reply:
I think the answer to this question is the same as the one [above](https://github.com/rapidsai/cudf/pull/18084#discussion_r1974273926): the release callback for the exported arrays should not be doing anything other than decremented a shared pointer ref count to the same container. IOW both the arrow_table/arrow_column objects and any ArrowArray/ArrowDeviceArray/etc types we export to should only ever be "releasing" by deleting their shared pointer to the underlying data type. The actual data is only freed when the container is freed, which only happens when the ref count of the shared pointer to it goes to zero. Nothing gets direct access to the container instance, it's always mediated by a shared_ptr.